File: ui/src/api/repository.ts
const API_BASE = '/api/repository'

export interface AssetMetadata {
  id: string
  type: string
  name: string
  description: string
  version: string
  category: string
  tags: string[]
  author?: string
  rating: number
  usage_count: number
  dependencies: string[]
  config_schema: Record<string, unknown>
}

export interface SearchResult {
  asset: AssetMetadata
  score: number
  highlights: Record<string, string[]>
}

export interface SearchResponse {
  success: boolean
  results: SearchResult[]
  total: number
  query?: string
  took_ms: number
}

export interface RepositoryStats {
  total_assets: number
  assets_by_type: Record<string, number>
  assets_by_category: Record<string, number>
  top_tags: Array<{ tag: string; count: number }>
  recent_assets: AssetMetadata[]
  popular_assets: AssetMetadata[]
}

export interface Skill {
  id: string
  name: string
  description: string
  category: string
  version: string
  system_prompt: string
  parameters: Array<{
    name: string
    type: string
    description: string
    required: boolean
    default?: unknown
  }>
  required_tools: string[]
  tags: string[]
  rating: number
}

export async function searchRepository(params: {
  query?: string
  types?: string[]
  categories?: string[]
  tags?: string[]
  sort_by?: string
  limit?: number
  offset?: number
}): Promise<SearchResponse> {
  const searchParams = new URLSearchParams()
  if (params.query) searchParams.set('query', params.query)
  if (params.types) params.types.forEach(t => searchParams.append('types', t))
  if (params.categories) params.categories.forEach(c => searchParams.append('categories', c))
  if (params.tags) params.tags.forEach(t => searchParams.append('tags', t))
  if (params.sort_by) searchParams.set('sort_by', params.sort_by)
  if (params.limit) searchParams.set('limit', params.limit.toString())
  if (params.offset) searchParams.set('offset', params.offset.toString())

  const response = await fetch(`${API_BASE}/search?${searchParams}`)
  return response.json()
}

export async function getStats(): Promise<RepositoryStats> {
  const response = await fetch(`${API_BASE}/stats`)
  return response.json()
}

export async function getCategories(type?: string): Promise<string[]> {
  const url = type ? `${API_BASE}/categories?type=${type}` : `${API_BASE}/categories`
  const response = await fetch(url)
  return response.json()
}

export async function getTags(type?: string): Promise<string[]> {
  const url = type ? `${API_BASE}/tags?type=${type}` : `${API_BASE}/tags`
  const response = await fetch(url)
  return response.json()
}

export async function getAsset(type: string, id: string): Promise<AssetMetadata> {
  const response = await fetch(`${API_BASE}/assets/${type}/${id}`)
  return response.json()
}

export async function getSkills(category?: string): Promise<{ skills: Skill[]; total: number }> {
  const url = category ? `${API_BASE}/skills?category=${category}` : `${API_BASE}/skills`
  const response = await fetch(url)
  return response.json()
}

export async function getSkill(id: string): Promise<Skill> {
  const response = await fetch(`${API_BASE}/skills/${id}`)
  return response.json()
}

File: app/models/repository.py
"""
Repository models for agent/tool/skill discovery and search.
"""
from enum import Enum
from typing import Optional, List, Dict, Any, Union
from pydantic import BaseModel, Field
from datetime import datetime


class AssetType(str, Enum):
    """Types of assets in the repository."""
    AGENT = "agent"
    TOOL = "tool"
    SKILL = "skill"
    GRAPH = "graph"
    ADAPTER = "adapter"


class AssetVisibility(str, Enum):
    """Visibility of assets."""
    PUBLIC = "public"
    PRIVATE = "private"
    ORGANIZATION = "organization"


class AssetMetadata(BaseModel):
    """Common metadata for all repository assets."""
    id: str = Field(..., description="Unique asset identifier")
    type: AssetType = Field(..., description="Asset type")
    name: str = Field(..., description="Asset name")
    description: str = Field(default="", description="Asset description")
    version: str = Field(default="1.0.0")
    
    # Categorization
    category: str = Field(default="general")
    tags: List[str] = Field(default_factory=list)
    
    # Authorship
    author: Optional[str] = Field(default=None)
    organization: Optional[str] = Field(default=None)
    
    # Visibility and access
    visibility: AssetVisibility = Field(default=AssetVisibility.PUBLIC)
    
    # Statistics
    rating: float = Field(default=0.0, ge=0.0, le=5.0)
    rating_count: int = Field(default=0)
    usage_count: int = Field(default=0)
    download_count: int = Field(default=0)
    
    # Timestamps
    created_at: Optional[datetime] = Field(default=None)
    updated_at: Optional[datetime] = Field(default=None)
    
    # Dependencies
    dependencies: List[str] = Field(default_factory=list, description="IDs of required assets")
    
    # Configuration schema
    config_schema: Dict[str, Any] = Field(default_factory=dict, description="JSON schema for configuration")
    
    # Documentation
    readme: Optional[str] = Field(default=None, description="Markdown documentation")
    examples: List[Dict[str, Any]] = Field(default_factory=list)
    
    class Config:
        extra = "allow"


class SearchFilter(BaseModel):
    """Filters for repository search."""
    types: Optional[List[AssetType]] = Field(default=None, description="Filter by asset types")
    categories: Optional[List[str]] = Field(default=None, description="Filter by categories")
    tags: Optional[List[str]] = Field(default=None, description="Filter by tags")
    author: Optional[str] = Field(default=None, description="Filter by author")
    organization: Optional[str] = Field(default=None, description="Filter by organization")
    visibility: Optional[AssetVisibility] = Field(default=None)
    min_rating: Optional[float] = Field(default=None, ge=0.0, le=5.0)
    created_after: Optional[datetime] = Field(default=None)
    created_before: Optional[datetime] = Field(default=None)


class SearchRequest(BaseModel):
    """Request for searching the repository."""
    query: Optional[str] = Field(default=None, description="Search query string")
    filters: Optional[SearchFilter] = Field(default=None)
    sort_by: str = Field(default="relevance", description="Sort by: relevance, rating, usage, created, name")
    sort_order: str = Field(default="desc", description="Sort order: asc, desc")
    limit: int = Field(default=20, ge=1, le=100)
    offset: int = Field(default=0, ge=0)


class SearchResult(BaseModel):
    """A single search result."""
    asset: AssetMetadata
    score: float = Field(default=0.0, description="Relevance score")
    highlights: Dict[str, List[str]] = Field(default_factory=dict, description="Search highlights")


class SearchResponse(BaseModel):
    """Response for repository search."""
    success: bool = Field(default=True)
    results: List[SearchResult] = Field(default_factory=list)
    total: int = Field(default=0)
    query: Optional[str] = Field(default=None)
    took_ms: float = Field(default=0.0)


class AssetRating(BaseModel):
    """Rating for an asset."""
    asset_id: str
    user_id: str
    rating: float = Field(..., ge=1.0, le=5.0)
    review: Optional[str] = Field(default=None)
    created_at: datetime = Field(default_factory=datetime.utcnow)


class RepositoryStats(BaseModel):
    """Statistics for the repository."""
    total_assets: int = Field(default=0)
    assets_by_type: Dict[str, int] = Field(default_factory=dict)
    assets_by_category: Dict[str, int] = Field(default_factory=dict)
    top_tags: List[Dict[str, Any]] = Field(default_factory=list)
    recent_assets: List[AssetMetadata] = Field(default_factory=list)
    popular_assets: List[AssetMetadata] = Field(default_factory=list)
